Code/Framework/AzToolsFramework/AzToolsFramework/Prefab/Undo/PrefabUndoDelete.cpp

```diff
@@ -0,0 +1,96 @@
+ * For complete copyright and license terms please see the LICENSE at the root of this distribution.
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+ *
+ */
+
+#include <AzCore/Interface/Interface.h>
+#include <AzToolsFramework/Prefab/Instance/InstanceToTemplateInterface.h>
+#include <AzToolsFramework/Prefab/PrefabDomUtils.h>
+#include <AzToolsFramework/Prefab/PrefabSystemComponentInterface.h>
+#include <AzToolsFramework/Prefab/Undo/PrefabUndo.h>
+#include <AzToolsFramework/Prefab/Undo/PrefabUndoDelete.h>
+#include <AzToolsFramework/Prefab/Undo/PrefabUndoUtils.h>
+
+namespace AzToolsFramework
+{
+    namespace Prefab
+    {
+        PrefabUndoDeleteEntity::PrefabUndoDeleteEntity(const AZStd::string& undoOperationName)
+            : PrefabUndoBase(undoOperationName)
+        {
+        }
+
+        void PrefabUndoDeleteEntity::Capture(
+            const AZStd::vector<AZStd::string>& entityAliasPathList,
+            const AZStd::vector<const AZ::Entity*> parentEntityList,
+            Instance& focusedInstance)
+        {
+            m_templateId = focusedInstance.GetTemplateId();
+            m_redoPatch.SetArray();
+            m_undoPatch.SetArray();
+
+            PrefabDom& focusedTempalteDom = m_prefabSystemComponentInterface->FindTemplateDom(m_templateId);
+
+            PrefabDomReference cachedOwningInstanceDom = focusedInstance.GetCachedInstanceDom();
+
+            // Removes entity DOMs.
+            for (const AZStd::string& entityAliasPath : entityAliasPathList)
+            {
+                if (entityAliasPath.empty())
+                {
+                    AZ_Error("Prefab", false, "PrefabUndoDeleteEntity::Capture - Entity alias path is empty. Skips removal.");
+                    continue;
+                }
+
+                PrefabUndoUtils::AppendRemovePatch(m_redoPatch, entityAliasPath);
+
+                PrefabDomPath entityDomPathInFocusedTemplate(entityAliasPath.c_str());
+                const PrefabDomValue* entityDomInFocusedTemplate = entityDomPathInFocusedTemplate.Get(focusedTempalteDom);
+                PrefabUndoUtils::AppendAddEntityPatch(m_undoPatch, *entityDomInFocusedTemplate, entityAliasPath);
+
+                // Preemptively updates the cached DOM to prevent reloading instance.
+                if (cachedOwningInstanceDom.has_value())
+                {
+                    PrefabUndoUtils::UpdateCachedOwningInstanceDomWithRemoval(cachedOwningInstanceDom, entityAliasPath);
+                }
+            }
+
+            // Updates parent entity DOMs.
+            for (const AZ::Entity* parentEntity : parentEntityList)
+            {
+                // This is actually a from-focused path because focused instance is the owning instance in the current workflow.
+                const AZStd::string parentEntityAliasPath = m_instanceToTemplateInterface->GenerateEntityAliasPath(parentEntity->GetId());
+                PrefabDomPath parentEntityAliasDomPathFromFocused(parentEntityAliasPath.c_str());
```

Comment:
I didn't follow. Do you mean "is there an API that can be used to pass in owningInstance"?

Reply:
Yeah, I'm wondering if we should replace m_instanceToTemplateInterface->GenerateEntityAliasPath with a call that we can pass focusedInstance to as an argument.

Reply:
I think it is not needed for now. All similar APIs in this class are taking in an entity id. fyi We recently added a new API: GenerateEntityPathFromFocusedPrefab(entityId).

I'm wondering if later it would be good to have single generic interface that handles:
1. AliasPath from owning instance
2. AliasPath from focused instance (if provided)

@srikappa-amzn any thoughts?

Reply:
`        void PrefabUndoDeleteEntity::Capture(
            const AZStd::vector<AZStd::string>& entityAliasPathList,
            const AZStd::vector<const AZ::Entity*> parentEntityList,
            Instance& focusedInstance)`

Since this function takes an instance reference as an argument, it seems confusing to later call other API methods that retrieve the "focused instance" on their own. My suggestion is to either not pass in focusedInstance at all and use API calls that know what the focused instance is, or pass in the focused instance but use API calls that also take a focused instance as an argument.

I may be missing something though, so let me know if I am. :)

Reply:
I think it's better to use GenerateEntityPathFromFocusedPrefab() in this use case rather than adding a comment to explain the intention. It partly addresses @michabr point about not being obvious where the path is generated from. It doesn't use the focused instance you already have in this function though. If you want to avoid that too, you can provide a new private helper function in this class that implements something like GenerateEntityPathFromFocusedPrefab itself. But I don't think it's worth the effort to do that here. Getting the focused instance twice shouldn't be an issue performance wise

Reply:
I'm more concerned about having the API be super intuitive rather than this being a performance issue, but I'll go ahead and approve the PR since this doesn't seem like a major concern.